src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -0,0 +1,802 @@
+                    Marshal.PtrToStructure(structPtr, status);
+
+                    if (string.Equals(_eitherName, status.serviceName, StringComparison.OrdinalIgnoreCase) ||
+                        string.Equals(_eitherName, status.displayName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (_name == null)
+                        {
+                            _name = status.serviceName;
+                        }
+
+                        if (_displayName == null)
+                        {
+                            _displayName = status.displayName;
+                        }
+
+                        _eitherName = string.Empty;
+                        return;
+                    }
+                }
+
+                throw new InvalidOperationException(SR.Format(SR.NoService, _eitherName, _machineName));
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(memory);
+                if (databaseHandle != IntPtr.Zero)
+                {
+                    Interop.mincore.CloseServiceHandle(databaseHandle);
+                }
+            }
+        }
+
+        private static IntPtr GetDataBaseHandleWithAccess(string machineName, int serviceControlManagerAccess)
+        {
+            IntPtr databaseHandle = IntPtr.Zero;
+            if (machineName.Equals(".") || machineName.Length == 0)
+            {
+                databaseHandle = Interop.mincore.OpenSCManager(null, null, serviceControlManagerAccess);
+            }
+            else
+            {
+                databaseHandle = Interop.mincore.OpenSCManager(machineName, null, serviceControlManagerAccess);
+            }
+
+            if (databaseHandle == IntPtr.Zero)
+            {
+                Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
+                throw new InvalidOperationException(SR.Format(SR.OpenSC, machineName), inner);
+            }
+
+            return databaseHandle;
+        }
+
+        private void GetDataBaseHandleWithConnectAccess()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(GetType().Name);
+            }
+
+            // get a handle to SCM with connect access and store it in serviceManagerHandle field.
+            if (_serviceManagerHandle == null)
+            {
+                _serviceManagerHandle = new SafeServiceHandle(GetDataBaseHandleWithAccess(_machineName, Interop.SC_MANAGER_CONNECT));
+            }
+        }
+
+        private static IntPtr GetDataBaseHandleWithEnumerateAccess(string machineName)
+        {
+            return GetDataBaseHandleWithAccess(machineName, Interop.SC_MANAGER_ENUMERATE_SERVICE);
+        }
+
+        /// Gets all the device-driver services on the local machine.
+        public static ServiceController[] GetDevices()
+        {
+            return GetDevices(".");
```

Comment:
This "." string is used several times throughout the code.  Consider putting it into a const with an appropriate name.